File: graph_algorithms/GiantPizza.java
import java.util.*;
import java.io.*;



public class GiantPizza {

    private static int m;

    public static int not(int a) {
        if(a > m) {
            return a-m;
        }

        return a+m;
    }

    public static void main(String[] args) {
        FastScanner sc = new FastScanner();
        int n = sc.nextInt();
        m = sc.nextInt();


        List<List<Integer>> graph = new ArrayList<>();
        List<List<Integer>> transposedGraph = new ArrayList<>();
        for(int i=0; i<=2*m; i++) {
            graph.add(new ArrayList<>());
            transposedGraph.add(new ArrayList<>());
        }

        for(int i=1; i<=n ; i++) {
            char a = sc.next().charAt(0);
            int b = sc.nextInt();
            char c = sc.next().charAt(0);
            int d = sc.nextInt();

            if(a == '-') {
                b = not(b);
            }
            if(c == '-') {
                d = not(d);
            }

            graph.get(not(b)).add(d);
            transposedGraph.get(d).add(not(b));

            graph.get(not(d)).add(b);
            transposedGraph.get(b).add(not(d));
        }

        Stack<Integer> order = new Stack<>();
        boolean visited[] = new boolean[2*m+1];
        for(int i=1; i<=2*m; i++) {
            if(!visited[i]) {
                dfs(graph, visited, order, i);
            }
        }

        int componentId[] = new int[2*m+1];
        int count = 1;
        while(!order.isEmpty()) {
            int element = order.pop();

            if(componentId[element] != 0) {
                continue;
            }
            dfs2(transposedGraph, componentId, element, count);
            count++;
        }

        for(int i=1; i<=m; i++) {
            if(componentId[i] == componentId[not(i)]) {
                System.out.println("IMPOSSIBLE");
                return;
            }
        }

        StringBuilder s = new StringBuilder();
        for(int i=1; i<=m; i++) {
            if(componentId[i] > componentId[not(i)]) {
                s.append("+ ");
            } else {
                s.append("- ");
            }
        }
        System.out.println(s);

    }

    private static void dfs2(List<List<Integer>> graph, int componentId[], int node, int count) {
        if(componentId[node] != 0) {
            return;
        }

        componentId[node] = count;
        for(int neighbor : graph.get(node)) {
            dfs2(graph, componentId, neighbor, count);
        }
    }

    public static void dfs(List<List<Integer>> graph, boolean visited[], Stack<Integer> order, int node) {
        if(visited[node]) {
            return;
        }

        visited[node] = true;
        for(int neighbor : graph.get(node)) {
            dfs(graph, visited, order, neighbor);
        }
        order.add(node);
    }
    
}

class FastScanner {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer("");
 
    String next() {
        while (!st.hasMoreTokens())
            try {
                st = new StringTokenizer(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        return st.nextToken();
    }
 
    int nextInt() {
        return Integer.parseInt(next());
    }
 
    int[] readArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; i++)
            a[i] = nextInt();
        return a;
    }
 
    long nextLong() {
        return Long.parseLong(next());
    }
}

File: graph_algorithms/giant_pizza.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int m;

int not_op(int a) {
    return (a > m) ? a - m : a + m;
}

void dfs(const vector<vector<int>>& graph, vector<bool>& visited, stack<int>& order, int node) {
    if (visited[node]) return;
    
    visited[node] = true;
    for (int neighbor : graph[node]) {
        dfs(graph, visited, order, neighbor);
    }
    order.push(node);
}

void dfs2(const vector<vector<int>>& graph, vector<int>& componentId, int node, int count) {
    if (componentId[node] != 0) return;
    
    componentId[node] = count;
    for (int neighbor : graph[node]) {
        dfs2(graph, componentId, neighbor, count);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n >> m;

    vector<vector<int>> graph(2 * m + 1);
    vector<vector<int>> transposedGraph(2 * m + 1);

    for (int i = 1; i <= n; ++i) {
        char a, c;
        int b, d;
        cin >> a >> b >> c >> d;

        if (a == '-') b = not_op(b);
        if (c == '-') d = not_op(d);

        graph[not_op(b)].push_back(d);
        transposedGraph[d].push_back(not_op(b));

        graph[not_op(d)].push_back(b);
        transposedGraph[b].push_back(not_op(d));
    }

    stack<int> order;
    vector<bool> visited(2 * m + 1, false);
    
    for (int i = 1; i <= 2 * m; ++i) {
        if (!visited[i]) {
            dfs(graph, visited, order, i);
        }
    }

    vector<int> componentId(2 * m + 1, 0);
    int count = 1;
    
    while (!order.empty()) {
        int element = order.top();
        order.pop();

        if (componentId[element] == 0) {
            dfs2(transposedGraph, componentId, element, count++);
        }
    }

    for (int i = 1; i <= m; ++i) {
        if (componentId[i] == componentId[not_op(i)]) {
            cout << "IMPOSSIBLE\n";
            return 0;
        }
    }

    for (int i = 1; i <= m; ++i) {
        if (componentId[i] > componentId[not_op(i)]) {
            cout << "+ ";
        } else {
            cout << "- ";
        }
    }
    cout << endl;

    return 0;
}
